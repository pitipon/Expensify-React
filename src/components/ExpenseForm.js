import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();
console.log(now.format('MMM Do, YYYY'));

class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);

        let expense = props.expense;

        this.state = {
            description: expense ? expense.description : '',
            note: expense ? expense.note : '',
            amount: expense ? expense.amount : '',
            createdAt: expense ? moment(expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState({description});
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState({note});
    }

    onAmountChange = (e) => {
        const amount = e.target.value;

        // Check number with 2 digit .. ex 334.50
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState({amount});
        }
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            // set error state equal to 'Please provide description and amount'
            this.setState({ error: 'Please provide description and amount'});
        } else {
            
            console.log('Submit!!')
            // Trigger onSubmit to dispatch expense
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })

            // Reset value
            this.setState({ 
                description: '',
                amount: '',
                note: '',
                createdAt: moment(),
                error:'' 
            });
        }
    }

    render() {

        return (
            <div>
                {
                    this.state.error && (
                        <p>{this.state.error}</p>
                    )
                }
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={ createdAt => this.setState({ createdAt })} // PropTypes.func.isRequired
                        focused={this.state.calendarFocused} // PropTypes.bool
                        onFocusChange={({ focused }) => this.setState({ calendarFocused: focused })} // PropTypes.func.isRequired
                        numberOfMonths={1}
                        isOutsideRange={ day => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <button type='submit'>Submit</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm;