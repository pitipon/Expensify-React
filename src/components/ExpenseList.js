import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.map((expense) => (
                <ExpenseListItem {...expense} key={expense.id}/>            
            ))
        }
    </div>
);

const mapStateToProps = (state) => {
    console.log(44, state)
    return {
        expenses: selectExpenses(state.expenses, state.filter)
    }
}


// Enchance component with redux
const ConnectedExpenseList = connect(
    mapStateToProps
)(ExpenseList);

export default ConnectedExpenseList;