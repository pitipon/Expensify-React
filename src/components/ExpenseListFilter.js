import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

class ExpenseListFilter extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            calendarFocused: null
        }
    }

    onDatesChange = ({ startDate, endDate }) => {
        console.log(999, startDate, endDate)
        // Dispatch start Date
        this.props.dispatch(
            setStartDate(startDate)
        )

        // Dispatch end Date
        this.props.dispatch(
            setEndDate(endDate)
        )

    }

    render() {

        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filter.text} 
                    onChange={(e) => {
                        this.props.dispatch(setTextFilter(e.target.value))
                    }}
                />
                <select
                    value={this.props.filter.sortBy}
                    onChange={(e) => {
                        if (e.target.value === 'date') {
                            this.props.dispatch(sortByDate());
                        } else if (e.target.value === 'amount') {
                            this.props.dispatch(sortByAmount());
                        }
                    }}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filter.startDate} // momentPropTypes.momentObj or null,
                    startDateId="calendar_start_date" // PropTypes.string.isRequired,
                    endDate={this.props.filter.endDate} // momentPropTypes.momentObj or null,
                    endDateId="calendar_end_date" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={calendarFocused => this.setState({ calendarFocused })} // PropTypes.func.isRequired,
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                />
            </div>
        )
    }
}


const mapStateToProps = (state) => {
    return {
        filter: state.filter
    }
}

// Enchance component with redux
const ConnectedExpenseListFilter = connect(
    mapStateToProps
)(ExpenseListFilter);

export default ConnectedExpenseListFilter;

