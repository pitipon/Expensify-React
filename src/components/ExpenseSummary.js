import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({ expenseCount, expensesTotal}) => {
    const wordExpensesTotal = numeral(expensesTotal).format('$0,0.00');
    return (
        <div>
            <h1>Viewing {expenseCount} expense total {wordExpensesTotal}</h1>
        </div>
    )
}

const mapStateToProps = (state) => {
    // console.log(99999, state)
    const visibleExpenses = selectExpenses(state.expenses, state.filter);
    
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    }

}

const ConnectExpenseSummary = connect(mapStateToProps)(ExpenseSummary);

export default ConnectExpenseSummary;