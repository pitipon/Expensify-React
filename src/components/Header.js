import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <div>
            <NavLink exact to="/" activeClassName="is-active">Dashboard</NavLink>
            <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
            <NavLink to="/help" activeClassName="is-active">Help</NavLink>
            <button onClick={startLogout}>Logout</button>
        </div>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})

const ConnectHeader = connect(undefined, mapDispatchToProps)(Header);

export default ConnectHeader;