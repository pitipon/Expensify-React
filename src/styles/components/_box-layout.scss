.box-layout {
    align-items: center;
    // background: url('/images/bg.jpg');
    background-image: linear-gradient(-225deg, rgba(106, 0, 168, 0.6) 0%, rgba(16, 0, 61, 0.6) 50%), url('https://kitt.lewagon.com/placeholder/cities/berlin');
    background-size: cover;
    display: flex;
    height: 100vh;
    justify-content: center;
    width: 100vw;
  }
  
  .box-layout__box {
    background: fade-out(rgb(236, 236, 236), .1);
    border-radius: 5px;
    padding: $xl-size $xl-size;
    text-align: center;
    width: 40rem;
  }
  
  .box-layout__title {
    margin: 0 0 $m-size 0;
    line-height: 1;
  }
  
  .icon {
      width: 100px;
      height: 100px;
  }