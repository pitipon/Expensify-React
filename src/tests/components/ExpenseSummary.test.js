import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should correctly render ExpensesSummary with zero expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={0} expensesTotal={0}/>);
    expect(wrapper).toMatchSnapshot();
})

test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={235}/>);
    expect(wrapper).toMatchSnapshot();
})

test('should correctly render ExpensesSummary with multiple expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={3} expensesTotal={235333}/>);
    expect(wrapper).toMatchSnapshot();
})