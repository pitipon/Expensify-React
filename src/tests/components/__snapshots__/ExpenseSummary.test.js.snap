// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     expense total 
    $235.00
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expense 1`] = `
<div>
  <h1>
    Viewing 
    3
     expense total 
    $235,333.00
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with zero expense 1`] = `
<div>
  <h1>
    Viewing 
    0
     expense total 
    $0.00
  </h1>
</div>
`;
